### DISCUSSION TITLE
Scripting: events API for logging GUI interactions and changes to mindmap in real-time

### DISCUSSION THREAD

**Discussion by hi2u** [freeplane/freeplane#1275] [Help]
I want to do some scripting, that basically listens for any type of "event" in the GUI where I'm navigating / modifying my mindmap.  

I'd write my script to listen for these events, and then write some detailed info to some `.json` logs or something like that.  Initially I thought maybe I'd send them immediately as HTTP API requests, but I guess in that case it might miss things if the API is down.  So just saving to some detailed logs is probably simpler to start with at least, and solves that downtime issue.

The types of events I want to record in real-time:

* When I create a new node
* Delete a node
* Move a node
* Change the style of a node
* Focus a node (by clicking on it, or using keyboard arrows)
* Expand/collapse a node

When any of these actions are performed, my script would then log some metadata about the action that was taken, the metadata including:

* The filepath to the `.mm` file
* The relevant node ID
* The type of event that occurred
* Info about the event, e.g. when changing the style, the name/id of the new style being applied

Or otherwise, could this be achieved just by somehow cranking up Freeplane's own logging?  Is there a way to get this kind of low-level interaction detail in those logs?

[I had a look through the scripting docs.](https://docs.freeplane.org/scripting/Scripting.html)  But couldn't find anything about "events" or something that sounded similar.

1. Is this possible?  
2. Any links to docs that I could find would be awesome
3. Also any links to any example scripts that do something along the lines would be really great

**Comment by euu2021** [comment]
I don't know the answer to your question, but I can recommend those two reads:

- https://github.com/freeplane/freeplane/discussions/520#discussioncomment-3311095

- #544

**Comment by macmarrum** [comment]
The [Scripting API](https://docs.freeplane.org/api/) is limited as to what changes can be listened to. It exposes [NodeChangeListerner](https://docs.freeplane.org/api/org/freeplane/api/NodeChangeListener.html).
Here's an example of how it can be used: [MacmarrumNodeChangeListenerForModified.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/lib/MacmarrumNodeChangeListenerForModified.groovy)

Any change listener needs to be activated. You can use `<user-dir>/scripts/init/` to place a script, which will be executed on Freeplane's start, so it's fairly easy to activate such a listener per Freeplane instance. But if this is to be done per mind map, you need to activate your listener each time you open the mind map. Fortunately, there is an add-on which automates this: [ScriptExecutionOnMindmapOpen.addon.mm](https://github.com/macmarrum/freeplane-ScriptExecutionOnMindmapOpen/releases)

Other events are considered by other listeners, which are available in the Java code, i.e. in Freeplane internal API. One of them is **MapChangeListener**. Here's an example of how I use it: [MacmarrumChangeListenerUtils.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/lib/MacmarrumChangeListenerUtils.groovy)
The problem with using Freeplane's internal API (as opposed to the Scripting API) is that when it changes, the script will stop working.

I know that focus and fold/unfold (expand/collapse) is view-specific (see the menu item `View->New map view`), therefore most probably in some other responsibility than NodeChangeListerner or MapChangeListener.

I'd be interested in your findings.

----

Here's how to install **MacmarrumNodeChangeListenerForModified** and enable it:

1. Download [MacmarrumNodeChangeListenerForModified.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/lib/MacmarrumNodeChangeListenerForModified.groovy) and save it to `<user-dir>/lib`
_Note: You can use `Tools->Open user directory`_
2. Download and install the latest release of [ScriptExecutionOnMindmapOpen.addon.mm](https://github.com/macmarrum/freeplane-ScriptExecutionOnMindmapOpen/releases)
_Note: simply open the .mm file and agree to install the add-on_
3. In your mindmap, in root node, add an attribute `scriptOnMapOpen` with the value `node.mindMap.addListener(new MacmarrumNodeChangeListenerForModified())`
4. Set your `Preferences…->Behaviour->Data formatting and parsing->Standard date-time format`, e.g. `yyyy-MM-dd HH:mm:ss`
5. Make sure you have `Preferences…->Plugins->Scripting->Script execution enabled`
6. Save your mindmap and restart Freeplane
7. Change the text, note or icons of a node and verify that the current date and time (in your date-time format) was inserted in an attribute called `Modified`
8. If you want the name of the attribute to be different, change it in `MacmarrumNodeChangeListenerForModified.groovy` and restart Freeplane.

----

Here's how to install **MacmarrumChangeListenerUtils.groovy** and enable the ID_ON_CREATE feature:

1. Download [MacmarrumChangeListenerUtils.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/lib/MacmarrumChangeListenerUtils.groovy) and place it in `<user-dir>/lib`.
_Note: You can use `Tools->Open user directory`_
2. Download and install the latest release of [ScriptExecutionOnMindmapOpen.addon.mm](https://github.com/macmarrum/freeplane-ScriptExecutionOnMindmapOpen/releases)
_Note: simply open the .mm file and agree to install the add-on_
3. In the root node of your mindmap
   1. Add an attribute named `scriptOnMapOpen` with the value `MacmarrumChangeListenerUtils.toggleMapChangeListener(node)`
   2. Add an attribute named `MacmarrumFeatures` with the value `ID_ON_CREATE`
   3. Add an attribute named `IdLocation` with the value `CORE` or `DETAILS` or `NOTE` depending where you want ID_NUMBER to be located
4. Restart your Freeplane
